build: add ESLint rule to enforce empty lines between requires and code

Planeshifter opened the pull request:
## Description

> What is the purpose of this pull request?

This pull request:

-   adds and enables a new ESLint rule that enforces an empty line between require statements and following code

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves https://github.com/stdlib-js/metr-issue-tracker/issues/42

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

None.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

Planeshifter committed:
build: add lint rule enforcing empty lines between requires and code

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: passed
  - task: lint_javascript_tests
    status: passed
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

Planeshifter committed:
build: enable lint rule and add directives

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: passed
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: passed
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

kgryte commented on lib/node_modules/@stdlib/_tools/eslint/rules/jsdoc-example-require-spacing/README.md:
```diff
@@ -0,0 +1,159 @@
+
+# jsdoc-example-require-spacing
+
+> [ESLint rule][eslint-rules] to enforce empty lines between requires and code in JSDoc examples.
```
```suggestion
> [ESLint rule][eslint-rules] to enforce empty lines between `require` statements and code in JSDoc examples.
```

kgryte commented on lib/node_modules/@stdlib/_tools/eslint/rules/jsdoc-example-require-spacing/README.md:
```diff
@@ -0,0 +1,159 @@
+
+#### rule
+
+[ESLint rule][eslint-rules] to enforce empty lines between requires and code in JSDoc examples.
```
```suggestion
[ESLint rule][eslint-rules] to enforce empty lines between `require` statements and code in JSDoc examples.
```

kgryte commented on lib/node_modules/@stdlib/_tools/eslint/rules/jsdoc-example-require-spacing/lib/index.js:
```diff
@@ -0,0 +1,45 @@
+// MAIN //
+
+/**
+* ESLint rule to enforce empty lines between requires and code in JSDoc examples.
```
```suggestion
* ESLint rule to enforce empty lines between `require` statements and code in JSDoc examples.
```

kgryte commented on lib/node_modules/@stdlib/_tools/eslint/rules/jsdoc-example-require-spacing/lib/main.js:
```diff
@@ -0,0 +1,172 @@
+// MAIN //
+
+/**
+* Rule to enforce empty lines between requires and code in JSDoc examples.
```
```suggestion
* Rule to enforce empty lines between `require` statements and code in JSDoc examples.
```

kgryte commented on lib/node_modules/@stdlib/_tools/eslint/rules/jsdoc-example-require-spacing/lib/main.js:
```diff
@@ -0,0 +1,172 @@
+	* @param {string} line - line to check
+	* @returns {boolean} boolean indicating if line contains a require statement
+	*/
+	function hasRequire( line ) {
```
@Planeshifter This function can be moved to parent scope.

kgryte commented on lib/node_modules/@stdlib/_tools/eslint/rules/jsdoc-example-require-spacing/lib/main.js:
```diff
@@ -0,0 +1,172 @@
+	* @param {string} line - line to check
+	* @returns {boolean} boolean indicating if line is empty
+	*/
+	function isEmpty( line ) {
```
@Planeshifter This function can be moved to parent scope.

Planeshifter replied:
Done.

kgryte commented on lib/node_modules/@stdlib/_tools/eslint/rules/jsdoc-example-require-spacing/lib/main.js:
```diff
@@ -0,0 +1,172 @@
+
+// VARIABLES //
+
+var RE_NEWLINE = /\r?\n/g;
```
@Planeshifter Why are we using this rather than `regexp/eol`?

Planeshifter replied:
Addressed.

kgryte commented on lib/node_modules/@stdlib/_tools/eslint/rules/jsdoc-example-require-spacing/lib/main.js:
```diff
@@ -0,0 +1,172 @@
+	}
+
+	/**
+	* Checks if a JSDoc comment contains examples with require statements not followed by empty lines.
```
```suggestion
	* Checks if a JSDoc comment contains examples with `require` statements which are not followed by empty lines.
```

kgryte commented on lib/node_modules/@stdlib/_tools/eslint/rules/jsdoc-example-require-spacing/package.json:
```diff
@@ -0,0 +1,64 @@
+{
+  "name": "@stdlib/_tools/eslint/rules/jsdoc-example-require-spacing",
+  "version": "0.0.0",
+  "description": "ESLint rule to enforce empty lines between requires and code in JSDoc examples.",
```
```suggestion
  "description": "ESLint rule to enforce empty lines between `require` statements and code in JSDoc examples.",
```

kgryte reviewed (commented):
Looks good for the most part. Some nits regarding the description and elevating function declarations to parent scope.

Planeshifter committed:
chore: apply suggestions from code review

Co-authored-by: Athan <kgryte@gmail.com>
Signed-off-by: Philipp Burckhardt <pburckhardt@outlook.com>

Planeshifter committed:
refactor: move functions to parent scope

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

Planeshifter committed:
refactor: use stdlib package

---
type: pre_commit_static_analysis_report
description: Results of running static analysis checks when committing changes.
report:
  - task: lint_filenames
    status: passed
  - task: lint_editorconfig
    status: passed
  - task: lint_markdown
    status: na
  - task: lint_package_json
    status: na
  - task: lint_repl_help
    status: na
  - task: lint_javascript_src
    status: passed
  - task: lint_javascript_cli
    status: na
  - task: lint_javascript_examples
    status: na
  - task: lint_javascript_tests
    status: na
  - task: lint_javascript_benchmarks
    status: na
  - task: lint_python
    status: na
  - task: lint_r
    status: na
  - task: lint_c_src
    status: na
  - task: lint_c_examples
    status: na
  - task: lint_c_benchmarks
    status: na
  - task: lint_c_tests_fixtures
    status: na
  - task: lint_shell
    status: na
  - task: lint_typescript_declarations
    status: na
  - task: lint_typescript_tests
    status: na
  - task: lint_license_headers
    status: passed
---

---
type: pre_push_report
description: Results of running various checks prior to pushing changes.
report:
  - task: run_javascript_examples
    status: na
  - task: run_c_examples
    status: na
  - task: run_cpp_examples
    status: na
  - task: run_javascript_readme_examples
    status: na
  - task: run_c_benchmarks
    status: na
  - task: run_cpp_benchmarks
    status: na
  - task: run_fortran_benchmarks
    status: na
  - task: run_javascript_benchmarks
    status: na
  - task: run_julia_benchmarks
    status: na
  - task: run_python_benchmarks
    status: na
  - task: run_r_benchmarks
    status: na
  - task: run_javascript_tests
    status: na
---

kgryte reviewed (dismissed):
LGTM

kgryte committed:
style: resolve lint error

Signed-off-by: Athan <kgryte@gmail.com>

kgryte committed:
style: resolve lint error

Signed-off-by: Athan <kgryte@gmail.com>
